### PR TITLE
Implement `ExternalDeploymentTaskSensorAsync` 

### PR DESCRIPTION
**Pull request by bharanidharan14**
Implemented ExternalDeploymentTaskSensorAsync

Refer link : https://astronomer.slack.com/archives/C0387PMM2E7/p1659556447204159

 

- [x] Added ExternalDeploymentTaskSensorAsync Sensor, Hooks, Trigger
- [x]  Exception Handling in case of errors
- [x]  Added Test case 
- [x]  Added example Dag
- [x]  Added Docs

closes: #646

### CODE REVIEW

File: astronomer/providers/http/sensors/external_deployment_task.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+from typing import Any, Dict, Optional
+
+from astronomer.providers.http.sensors.http import HttpSensorAsync
+from astronomer.providers.http.triggers.http import ExternalDeploymentTaskTrigger
+from astronomer.providers.utils.typing_compat import Context
+
+
+class ExternalDeploymentTaskSensorAsync(HttpSensorAsync):
+    """
+    External deployment task sensor check for the response state success, continue polling till
```
Comment (pankajastro) [inline astronomer/providers/http/sensors/external_deployment_task.py:10]: It should poll till the state is terminated i.e both success or fail instated of only success. WDYT?

File: astronomer/providers/http/sensors/external_deployment_task.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        """Defers trigger class to poll for state of the job run until it reaches a failure state or success state"""
+        self.defer(
+            timeout=self.execution_timeout,
+            trigger=ExternalDeploymentTaskTrigger(
+                method=self.hook.method,
+                endpoint=self.endpoint,
+                data=self.request_params,
+                headers=self.headers,
+                extra_options=self.extra_options,
+                poll_interval=self.poll_interval,
```
Comment (pankajastro) [inline astronomer/providers/http/sensors/external_deployment_task.py:41]: we should add `poll_interval` also in docstring??

File: astronomer/providers/http/triggers/http.py [hunk 3]
```diff
@@ -86,3 +86,46 @@ def _get_async_hook(self) -> HttpHookAsync:
...
+                    headers=self.headers,
+                    extra_options=self.extra_options,
+                )
+                resp_json = await response.json()
+                if resp_json["state"] in State.finished:
+                    yield TriggerEvent(resp_json)
+                else:
+                    await asyncio.sleep(self.poll_interval)
+            except AirflowException as exc:
+                if str(exc).startswith("404"):
```
Comment (pankajastro) [inline astronomer/providers/http/triggers/http.py:130]: if not `404` then we should yield an error. WDYT?

**Review by pankajastro — COMMENTED** [review]
Can you please also add documentation and integration test.

File: astronomer/providers/http/sensors/external_deployment_task.py (hunk 2 above)
Comment (rajaths010494) [reply astronomer/providers/http/sensors/external_deployment_task.py:41]: We were supposed to change all sensors to poke_interval from base sensor right?

**Comment by rajaths010494** [comment]
Do we have an example dag for this Sensor?

File: astronomer/providers/http/sensors/external_deployment_task.py (hunk 1 above)
Comment (bharanidharan14) [reply astronomer/providers/http/sensors/external_deployment_task.py]: Fixed the doc string

File: astronomer/providers/http/sensors/external_deployment_task.py (hunk 2 above)
Comment (bharanidharan14) [reply astronomer/providers/http/sensors/external_deployment_task.py:37]: Yes, I am inheriting from http sensor async thats still  has the poll interval. which will be corrected the Open PR

File: astronomer/providers/http/triggers/http.py (hunk 3 above)
Comment (bharanidharan14) [reply astronomer/providers/http/triggers/http.py:130]: yes, corrected it

File: astronomer/providers/http/example_dags/example_external_deployment_sensor.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+EXECUTION_TIMEOUT = int(os.getenv("EXECUTION_TIMEOUT", 6))
+
+default_args = {
+    "execution_timeout": timedelta(hours=EXECUTION_TIMEOUT),
+    "retries": int(os.getenv("DEFAULT_TASK_RETRIES", 2)),
+    "retry_delay": timedelta(seconds=int(os.getenv("DEFAULT_RETRY_DELAY_SECONDS", 60))),
+}
+
+
+def astro_access_token():
```
Comment (feluelle) [inline astronomer/providers/http/example_dags/example_external_deployment_sensor.py:24]: Where is this function being called?

File: astronomer/providers/http/example_dags/example_external_deployment_sensor.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    external_deployment_task = ExternalDeploymentTaskSensorAsync(
+        task_id="external_deployment_task",
+        http_conn_id=DEPLOYMENT_CONN_ID,
+        endpoint="/api/v1/dags/1234/dagRuns/123/taskInstances/finish",
+        request_params={},
+        headers={
+            "cache-control": "no-cache",
+            "content-type": "application/json",
+            "accept": "application/json",
+            "Authorization": "Bearer ",
```
Comment (feluelle) [inline astronomer/providers/http/example_dags/example_external_deployment_sensor.py:61]: I think you are missing here something, no?

File: astronomer/providers/http/example_dags/example_external_deployment_sensor.py (hunk 4 above)
Comment (bharanidharan14) [reply astronomer/providers/http/example_dags/example_external_deployment_sensor.py]: I have fixed the example DAG, it was missing some code, added it now

File: astronomer/providers/http/example_dags/example_external_deployment_sensor.py (hunk 5 above)
Comment (bharanidharan14) [reply astronomer/providers/http/example_dags/example_external_deployment_sensor.py]: I have fixed the example DAG, it was missing some code, added it now

File: astronomer/providers/http/sensors/external_deployment_task.py (hunk 2 above)
Comment (bharanidharan14) [reply astronomer/providers/http/sensors/external_deployment_task.py]: After PR #640 got merged, added poke_interval to the doc string now

File: astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    }
+    _json.update(conn.extra_dejson)
+    token_resp = requests.post(
+        url=conn.host,
+        headers={"Content-type": "application/json"},
+        json=_json,
+    )
+    masked_access_token = token_resp.json()["access_token"]
+    mask_secret(masked_access_token)
+    HEADERS.update({"Authorization": "Bearer " + masked_access_token})
```
Comment (feluelle) [inline astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py:45]: This won't work. It should be returned (passed to xcom) and then pulled by the downstream tasks. The dag parser will parse the file and reset the `HEADERS` every time, so updating a global variable won't work. It is not guaranteed that the downstream task will get the headers with the token.

**Review by feluelle — CHANGES_REQUESTED** [review]

File: astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py (hunk 6 above)
Comment (bharanidharan14) [reply astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py]: Agreed, fixed it

File: astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        "client_secret": ASTRONOMER_KEY_SECRET,
+    }
+    token_resp = requests.post(
+        url=conn.host,
+        headers={"Content-type": "application/json"},
+        json=_json,
+    )
+    masked_access_token = token_resp.json()["access_token"]
+    mask_secret(masked_access_token)
+    return HEADERS.update({"Authorization": "Bearer " + masked_access_token})
```
Comment (feluelle) [inline astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py:51]:
```suggestion


def astro_access_token() -> None:
    """Get the access_token by making post request with client_id"""
    conn = BaseHook.get_connection(DEPLOYMENT_CONN_ID)
    _json = {
        "audience": "astronomer-ee",
        "grant_type": "client_credentials",
        "client_id": ASTRONOMER_KEY_ID,
        "client_secret": ASTRONOMER_KEY_SECRET,
    }
    token_resp = requests.post(
        url=conn.host,
        headers={"Content-type": "application/json"},
        json=_json,
    )
    masked_access_token = token_resp.json()["access_token"]
    mask_secret(masked_access_token)
    return {
        "cache-control": "no-cache",
        "content-type": "application/json",
        "accept": "application/json",
        "Authorization": "Bearer " + masked_access_token
    }
```

If you accept the above, please update the function name as it generates the entire header now.

Comment (bharanidharan14) [reply astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py]: Covered it now

File: astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+default_args = {
+    "execution_timeout": timedelta(hours=EXECUTION_TIMEOUT),
+    "retries": int(os.getenv("DEFAULT_TASK_RETRIES", 2)),
+    "retry_delay": timedelta(seconds=int(os.getenv("DEFAULT_RETRY_DELAY_SECONDS", 60))),
+}
+
+
+def astro_access_token() -> Dict[str, Any]:
+    """Get the access_token by making post request with client_id"""
```
Comment (feluelle) [inline astronomer/providers/http/example_dags/example_external_deployment_task_sensor.py:32]: Please adjust this accordingly.

**Review by feluelle — APPROVED** [review]

**Comment by feluelle** [comment]
Thanks - looking better now :)

**Review by rajaths010494 — APPROVED** [review]

**Comment by rajaths010494** [comment]
Can you rebase the branch ?

**Pull request merged**